### ISSUE TITLE
newrelic_rpm slows down requests *4x* when used with graphql dataloader

### ISSUE THREAD

**Issue by bmulholland** [newrelic/newrelic-ruby-agent#1345]
## Description

Following on to https://github.com/newrelic/newrelic-ruby-agent/issues/1264, I deployed v8.10.0 to production. Then I deployed, in isolation, `use GraphQL::Tracing::NewRelicTracing, set_transaction_name: true`. I found that, for key queries in our app, performance degraded from 1-2 seconds to 6-7s, about a 4x slowdown. This is better than 10x, but still unacceptable for production use.

## For Maintainers Only or Hero Triaging this bug
*Suggested Priority (P1,P2,P3,P4,P5): P5 (until more info is received)*
*Suggested T-Shirt size (S, M, L, XL, Unknown): Unknown*

**Comment by fallwith** [comment]
Hi @bmulholland,

Thanks very much for your efforts on driving this one forward and trying out the changes that went into v8.10.0. We're glad to see that there has been some improvement with that release.

One thing that would help us better understand and make progress on the performance problem would be a standalone method of reproduction. It would be great to have a gist or repo with operations that could be ran with and without the `NewRelicTracing` being enabled to demonstrate the performance difference. With such a code base, we could keep pointing it at different test builds of the New Relic Ruby agent until we see improvements.

If you yourself or anyone else from the Ruby GraphQL community has the time to put such an issue reproduction code base together, those efforts would be tremendously helpful and appreciated. If not, then no worries - we will still endeavor to better understand and work on reproducing the issue on our end.

**Comment by bmulholland** [comment]
And thanks for all your (and the team's) work! Apologies that I wasn't able to test this out pre-release.

Since this is blocking us from profiling requests, its prioritization will be driven by our need to investigate performance of a GraphQL query. Unfortunately, that means I can't justify investing time into a repro until a performance issue comes up. I wish I could help more! If you do investigate it further in the meantime, good luck; otherwise, I completely understand that you can't make progress without any way to look into it :)

**Comment by kaylareopelle** [comment]
Hi @bmulholland! 👋 

We have a [potential fix](https://github.com/newrelic/newrelic-ruby-agent/pull/1927) for a [different issue](https://github.com/newrelic/newrelic-ruby-agent/issues/1910) that may also improve the agent's performance when using Dataloader.

Could you test the fix and let us know how it impacts your app's performance?

To test, update your Gemfile's reference to `newrelic_rpm` to point at the fix's branch:
```
gem 'newrelic_rpm', git: 'https://github.com/newrelic/newrelic-ruby-agent/', branch: 'must_go_faster'
```
Then bundle, make sure `instrumentation.thread.tracing` set to `true` and run the problematic queries.

As a head's up, this fix points to the 9.x branch of our agent. If you haven't updated to 9.x yet, there may be other changes you need to make to get things working. [Here's the migration guide for 8.x to 9.x.](https://docs.newrelic.com/docs/apm/agents/ruby-agent/getting-started/migration-9x-guide/)

**Comment by bmulholland** [comment]
Hey, thanks for reporting in. We've decided to switch off of NewRelic for unrelated reasons, so we won't be able to test this. I'll close the ticket, hopefully the work already done has solved it for others :)

**Comment by kaylareopelle** [comment]
We had a New Relic internal user test this out the fix on a service that uses GraphQL Dataloader and they went from a > 5 seconds delay to a < 15ms delay. We have confidence this issue is resolved in version 9.2.0 of the agent.

**Comment by kaylareopelle** [comment]
For whoever may come across this issue: 

We had a New Relic internal user test this out the fix on a service that uses GraphQL Dataloader and they went from a > 5 seconds delay to a < 15ms delay. We have confidence this issue is resolved in version 9.2.2+ of the agent.
